### PR TITLE
Short event life option

### PR DESCRIPTION
**Pull request by stvnrlly**
This adds a variable in `.env`, `SCHEDULER_EXPIRATION_CHECK`. The default value of `normal` runs expiration checks as they currently exist. Setting the value to `frequent` adds an option for each agent to keep events for "A minute" and does expiration checks every 3 minutes.

This is useful for running high-volume agents on something like Heroku where the free database is limited.

### CODE REVIEW

**Comment by dsander** [comment]
I see your use case, however I have the feeling that this adds a lot of conditionals for a very special case.

How about we change the interval for `cleanup_expired_events!` to once an hour and add an hourly retention schedule? That would still allow about 150 events per second to be created on the free Heroku plan. You could then convert `EVENT_RETENTION_SCHEDULES` to minutes and avoid the conditionals.

@cantino What do you think about this?

**Comment by cantino** [comment]
@stvnrlly, what's the smallest cleanup period that would meet your needs?  Would every 6 hours work, or every 3?  Is the problem that you have a small database?

**Comment by stvnrlly** [comment]
One of the major reasons that I like frequent cleanup here is that it allows me to still see other events on the events page. With less frequent cleanup, nearly every event is from this one high-volume agent, making it difficult to check the output of other agents at a glance.

As for the actual limitations of the system, 6 hours is probably sufficient to keep under the limit, but it depends on how many of these types of agents I'm running and what the agents it feeds do. I can imagine having several of these real-time agent flows generating events, which decreases the total storage available to each. I've also added a peak detector to my setup, which also takes up a fair amount of memory even without generating events.

Perhaps a better approach is to have the `.env` option be the interval for cleanup rather than a preset. That way, I could set it to `3`, while somebody with lower volume could set it to `360` or just leave it as the default.

**Comment by stvnrlly** [comment]
Here's the scenario if you want to see what I'm actually doing here. (You might notice that the peak detector has a new option to set the initial wait period, which I haven't opened a PR about yet.)
```
{
  "name": "USA Analytics",
  "description": "No description provided",
  "source_url": false,
  "guid": "76ba563751b9fe13bd71122cdc015b7e",
  "tag_fg_color": "#ffffff",
  "tag_bg_color": "#5bc0de",
  "exported_at": "2015-04-29T14:06:53Z",
  "agents": [
    {
      "type": "Agents::PeakDetectorAgent",
      "name": "USA Analytics - Visitor Peak",
      "disabled": false,
      "guid": "92c225c63a3b7ba0aeddf59b3be09e4b",
      "options": {
        "expected_receive_period_in_days": "5",
        "group_by_path": "site.page",
        "value_path": "site.active_visitors",
        "message": "Hey! More people than usual ({{site.active_visitors}}) are on {{site.page}}",
        "window_duration_in_days": "30",
        "min_peak_spacing_in_days": "2",
        "std_multiple": "4",
        "wait_period": {
          "type": "days",
          "value": "7"
        }
      },
      "keep_events_for": 0,
      "propagate_immediately": true
    },
    {
      "type": "Agents::DataOutputAgent",
      "name": "USA Analytics - RSS Feed",
      "disabled": false,
      "guid": "d3ee1ddbe8c75d98c579fa5b324b25c8",
      "options": {
        "secrets": [
          "topfederalsites"
        ],
        "expected_receive_period_in_days": 2,
        "template": {
          "title": "Top Federal Sites",
          "description": "This is a feed of new entries to the Top 20 sites on analytics.usa.gov",
          "item": {
            "title": "{{ site.page }}",
            "description": "New site: {{ site.page_title }} with {{ site.active_visitors }} visitors",
            "link": "http://{{ site.page }}"
          },
          "link": "https://twitter.com/topfederalsites"
        }
      },
      "keep_events_for": 7,
      "propagate_immediately": true
    },
    {
      "type": "Agents::DeDuplicationAgent",
      "name": "USA Analytics - Deduplicator",
      "disabled": false,
      "guid": "1232d55c7ff619715dff4e44d56a5e7e",
      "options": {
        "property": "{{site.page}}",
        "lookback": "1000",
        "expected_update_period_in_days": "10"
      },
      "keep_events_for": 0,
      "propagate_immediately": true
    },
    {
      "type": "Agents::WebsiteAgent",
      "name": "USA Analytics - Live Feed",
      "disabled": false,
      "guid": "62c65edbaf04ef0d3be448fe633135b2",
      "options": {
        "expected_update_period_in_days": "2",
        "url": "https://analytics.usa.gov/data/live/top-pages-realtime.json",
        "type": "json",
        "mode": "all",
        "extract": {
          "site": {
            "path": "data[*]"
          }
        },
        "unzip": "gzip"
      },
      "schedule": "every_1m",
      "keep_events_for": 60,
      "propagate_immediately": true
    },
    {
      "type": "Agents::TwitterPublishAgent",
      "name": "USA Analytics - Tweet events",
      "disabled": false,
      "guid": "f790fd83d0bf3d8d19a9678644f2028d",
      "options": {
        "expected_update_period_in_days": "10",
        "message": "{{ message }}"
      },
      "keep_events_for": 0,
      "propagate_immediately": true
    },
    {
      "type": "Agents::EventFormattingAgent",
      "name": "USA Analytics - New Site Tweet Formatter",
      "disabled": false,
      "guid": "c36924f21b8c813734ce4140db6f6140",
      "options": {
        "instructions": {
          "message": "New site in the Top 20! Welcome to {{ site.page }} ({{ site.page_title | truncate: 60, 'â€¦' }}) with {{ site.active_visitors }} visitors"
        },
        "matchers": [

        ],
        "mode": "clean"
      },
      "keep_events_for": 0,
      "propagate_immediately": true
    }
  ],
  "links": [
    {
      "source": 0,
      "receiver": 4
    },
    {
      "source": 2,
      "receiver": 1
    },
    {
      "source": 2,
      "receiver": 5
    },
    {
      "source": 3,
      "receiver": 2
    },
    {
      "source": 3,
      "receiver": 0
    },
    {
      "source": 5,
      "receiver": 4
    }
  ]
}
```

**Comment by cantino** [comment]
I think it would be reasonable to just change the cleanup period to 3 hours for Huginn, and perhaps to add filtering options to the events page?

**Comment by stvnrlly** [comment]
Sure, that's fine. I'd just ask that the interval be set via a configurable `.env` option, to allow people to scale it in either direction.

**Comment by cantino** [comment]
That works for me.  Want to make that change?

**Comment by stvnrlly** [comment]
:+1:

**Comment by cantino** [comment]
@stvnrlly Just to clarify, are you making a PR that lowers the period on `cleanup_expired_events!` in `huginn_scheduler.rb` to a default of 3 hours, settable in .env, or would you like me to?  I wasn't sure if you were expecting me to or not, and I'm certainly happy to.

**Comment by stvnrlly** [comment]
Okay, I think that the last commit is what's needed to make it happen. I thought about adding some more options to `EVENT_RETENTION_SCHEDULES` in `agent.rb` when the cleanup schedule is set to lower values, but that seemed trickier as the current options don't differentiate between hours or days. If you want to take a crack at that (if you want to add those options), that would be helpful.

**Comment by stvnrlly** [comment]
Hmm, I just checked Travis's spec failure locally and can't reproduce it, so hopefully that's just a false positive.

File: .env.example [hunk 1]
```diff
@@ -146,6 +146,10 @@ ENABLE_SECOND_PRECISION_SCHEDULE=false
 # at the expense of time accuracy.
 #SCHEDULER_FREQUENCY=0.3
 
+# Specify the frequency with which the scheduler checks for event expiration.
+# You can use `m` for minutes, `h` for hours, and `d` for days.
+EVENT_EXPIRATION_CHECK="3h"
```
Comment (cantino) [inline .env.example]: I don't think we want the quotes on this.

File: lib/huginn_scheduler.rb [hunk 2]
```diff
@@ -114,7 +114,7 @@ def run!
     end
 
     # Schedule event cleanup.
-    @rufus_scheduler.cron "0 0 * * * " + tzinfo_friendly_timezone do
+    @rufus_scheduler.every ENV['EVENT_EXPIRATION_CHECK'] do
```
Comment (cantino) [inline lib/huginn_scheduler.rb]: Important to set a default, so something like `ENV['EVENT_EXPIRATION_CHECK'].presence || '6h'`

File: app/models/agent.rb [hunk 3]
```diff
@@ -129,12 +129,18 @@ def reload
   end
 
   def new_event_expiration_date
-    keep_events_for > 0 ? keep_events_for.days.from_now : nil
+    if keep_events_for == 60
```
Comment (cantino) [inline app/models/agent.rb]: How about something like this?

```ruby
class Agent < ActiveRecord::Base
  EVENT_RETENTION_SCHEDULES = [["Forever", 0], ['1 hour', 1.hour], ['6 hours', 6.hours], ["1 day", 1.day], *([2, 3, 4, 5, 7, 14, 21, 30, 45, 90, 180, 365].map {|n| ["#{n} days", n.days] })]

  def new_event_expiration_date
    keep_events_for > 0 ? keep_events_for.seconds.from_now : nil
  end

  def update_event_expirations!
    if keep_events_for == 0
      events.update_all :expires_at => nil
    else
      events.update_all "expires_at = " + rdbms_date_add("created_at", "SECOND", keep_events_for.to_i)
    end
  end
end
```

And then also add a migration that updates all Agent's `keep_events_for` values:

`db/migrate/20150507153436_update_keep_events_for_to_be_in_seconds.rb`

```ruby
class UpdateKeepEventsForToBeInSeconds < ActiveRecord::Migration
  class Agent < ActiveRecord::Base; end

  SECONDS_IN_DAY = 60 * 60 * 24

  def up
    Agent.update_all ['keep_events_for = keep_events_for * ?', SECONDS_IN_DAY]
  end

  def down
    Agent.update_all ['keep_events_for = keep_events_for / ?', SECONDS_IN_DAY]
  end
end
```

Comment (stvnrlly) [inline app/models/agent.rb]: After doing the migration, should I commit the changes to `db/schema.rb` as well?

Comment (cantino) [inline app/models/agent.rb]: Yes please!

**Comment by cantino** [comment]
This looks close @stvnrlly.  What's left to do?

**Comment by stvnrlly** [comment]
Unfortunately, it looks like the change to the cleanup time is causing a bunch of spec script failures. I fixed a few of the more obvious ones, but I wasn't able to tell why the others are failing.

**Comment by stvnrlly** [comment]
Here are some of the files that look like they still need updating:

- `spec/fixtures/agents.yml`
- `spec/models/agent_spec.rb`
- `spec/models/scenario_import_spec.rb`

I'm not sure which `keep_events_for` need to be updated to `x.days` and which ones should stay as-is, and changing them _all_ doesn't seem to resolve the spec failures. So, I'd appreciate if somebody else could take a look.

**Comment by cantino** [comment]
Okay, I'll take a look at this soon and see if I can see what else needs to be changed.

**Comment by cantino** [comment]
Sent you this PR that gets the specs green.

https://github.com/stvnrlly/huginn/pull/2

Once merged, we can do some manual QA.  This ended up complicated!

**Comment by cantino** [comment]
How did that look @stvnrlly?

**Comment by stvnrlly** [comment]
@cantino it looks good, as far as a I can tell, but I haven't had the opportunity to give it a test run yet :disappointed:. I'll be able to do that by this Sunday.

**Comment by cantino** [comment]
Sounds good.  You can merge it into your branch so that the final PR is
testable.

**Comment by stvnrlly** [comment]
Oh, yeah. I was thinking that I should test it myself, but it makes a lot more sense to let Travis take care of that.

**Comment by cantino** [comment]
@dsander Would you mind giving this a once-over?

File: lib/huginn_scheduler.rb [hunk 4]
```diff
@@ -114,7 +114,7 @@ def run!
     end
 
     # Schedule event cleanup.
-    @rufus_scheduler.cron "0 0 * * * " + tzinfo_friendly_timezone do
+    @rufus_scheduler.every ENV['EVENT_EXPIRATION_CHECK'].presence || '6h' do
```
Comment (dsander) [inline lib/huginn_scheduler.rb]: Shouldn't the fallback match the default in `.env.example`?

**Comment by dsander** [comment]
Apart from the minor inconsistency it looks good to me.

**Comment by cantino** [comment]
Okay, I have to do some manual testing on this, especially the Scenario converter, then we can get it out.

**Comment by cantino** [comment]
@stvnrlly Have you been using this successfully in your instance?

**Comment by stvnrlly** [comment]
I haven't been using that branch in production, but I can switch over and give it a shot.

**Comment by cantino** [comment]
I still need to manually test the Scenario code before this is mergable, but having you test the event handling would be good.

**Comment by stvnrlly** [comment]
@cantino I've been running this branch for a couple of days now. At first, I had to manually update a bunch of the Keep events for options, but that may be because I had previously been using my own hack to provide shorter times and therefore had non-standard values. I still have one agent giving me a `Validation failed: Keep events for is not included in the list` error and am having a hard time figuring out which one it is, but otherwise everything appears to be running normally.

**Comment by cantino** [comment]
Replaced by #923.

**Pull request closed without merging**